### ISSUE TITLE
Smarter cache for `snapshot_download`

### ISSUE THREAD

**Issue by patrickvonplaten** [huggingface/huggingface_hub#1342]
**Is your feature request related to a problem? Please describe.**

For `diffusers` many model repositories share identical checkpoint files. Currently when downloading two model repos that share many files both files are downloaded which is good since the whole model repository is self-contained, but arguable storage space is wasted.

Could we add features to `snapshot_download` that makes it "smart" in a way that it check whether a file has already been downloaded for another repo and then just creates a reference to this?

**Example**

Let's make the following example. We download both the v1-4 and v1-5 checkpoints from `diffusers`:

```py
from diffusers import DiffusionPipeline

DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4", cache_dir="local_cache")
DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", cache_dir="local_cache")
```

Now taking a look at:

```
tree local_cache
```

one gets:

```
├── models--CompVis--stable-diffusion-v1-4
│   ├── blobs
│   │   ├── 0e985ad52577b583221ab4bc6b8e275a9df6a2c6
│   │   ├── 0f998cd15e52f65656cd801497d171c5b9477cc7
│   │   ├── 193490b58ef62739077262e833bf091c66c29488058681ac25cf7df3d8190974
│   │   ├── 1a3e6d1a17dd6e722b4d4297023f979170219680
│   │   ├── 1b134cded8eb78b184aefb8805b6b572f36fa77b255c483665dda931fa0130c5
│   │   ├── 1d668d5d6799dea92e03bdd747de2365f8ef2393
│   │   ├── 2c2130b544c0c5a72d5d00da071ba130a9800fb2
│   │   ├── 469be27c5c010538f845f518c4f5e8574c78f7c8
│   │   ├── 5294955ff7801083f720b34b55d0f1f51313c5c5
│   │   ├── 5ba7bf706515bc60487ad0e1816b4929b82542d6
│   │   ├── 62d48b4d841a3178511fa453df0dae59b22089ace64609cc9d5353d0a7f37c26
│   │   ├── 76e821f1b6f0a9709293c3b6b51ed90980b3166b
│   │   ├── 770a47a9ffdcfda0b05506a7888ed714d06131d60267e6cf52765d61cf59fd67
│   │   ├── 93561f30a14e2a26eb3bf2ce3da74a52b36e90a3
│   │   ├── 97bed2ba03e1a77fe548fab801464c734b0a142b
│   │   └── df430a86cefe4115495e64c47936850adaeac0e9
│   ├── refs
│   │   └── main
│   └── snapshots
│       └── 3857c45b7d4e78b3ba0f39d4d7f50a2a05aa23d4
│           ├── feature_extractor
│           │   └── preprocessor_config.json -> ../../../blobs/5294955ff7801083f720b34b55d0f1f51313c5c5
│           ├── model_index.json -> ../../blobs/0e985ad52577b583221ab4bc6b8e275a9df6a2c6
│           ├── safety_checker
│           │   ├── config.json -> ../../../blobs/97bed2ba03e1a77fe548fab801464c734b0a142b
│           │   └── pytorch_model.bin -> ../../../blobs/193490b58ef62739077262e833bf091c66c29488058681ac25cf7df3d8190974
│           ├── scheduler
│           │   └── scheduler_config.json -> ../../../blobs/1a3e6d1a17dd6e722b4d4297023f979170219680
│           ├── text_encoder
│           │   ├── config.json -> ../../../blobs/df430a86cefe4115495e64c47936850adaeac0e9
│           │   └── pytorch_model.bin -> ../../../blobs/770a47a9ffdcfda0b05506a7888ed714d06131d60267e6cf52765d61cf59fd67
│           ├── tokenizer
│           │   ├── merges.txt -> ../../../blobs/76e821f1b6f0a9709293c3b6b51ed90980b3166b
│           │   ├── special_tokens_map.json -> ../../../blobs/2c2130b544c0c5a72d5d00da071ba130a9800fb2
│           │   ├── tokenizer_config.json -> ../../../blobs/5ba7bf706515bc60487ad0e1816b4929b82542d6
│           │   └── vocab.json -> ../../../blobs/469be27c5c010538f845f518c4f5e8574c78f7c8
│           ├── unet
│           │   ├── config.json -> ../../../blobs/0f998cd15e52f65656cd801497d171c5b9477cc7
│           │   └── diffusion_pytorch_model.bin -> ../../../blobs/62d48b4d841a3178511fa453df0dae59b22089ace64609cc9d5353d0a7f37c26
│           └── vae
│               ├── config.json -> ../../../blobs/1d668d5d6799dea92e03bdd747de2365f8ef2393
│               └── diffusion_pytorch_model.bin -> ../../../blobs/1b134cded8eb78b184aefb8805b6b572f36fa77b255c483665dda931fa0130c5
└── models--runwayml--stable-diffusion-v1-5
    ├── blobs
    │   ├── 19da7aaa4b880e59d56843f1fcb4dd9b599c28a1d9d9af7c1143057c8ffae9f1
    │   ├── 1a02ee8abc93e840ffbcb2d68b66ccbcb74b3ab3
    │   ├── 2c2130b544c0c5a72d5d00da071ba130a9800fb2
    │   ├── 469be27c5c010538f845f518c4f5e8574c78f7c8
    │   ├── 4d3e873ab5086ad989f407abd50fdce66db8d657
    │   ├── 5294955ff7801083f720b34b55d0f1f51313c5c5
    │   ├── 55d78924fee13e4220f24320127c5f16284e13b9
    │   ├── 5ba7bf706515bc60487ad0e1816b4929b82542d6
    │   ├── 5dbd88952e7e521aa665e5052e6db7def3641d03
    │   ├── 6866dceb3a870b077eb970ecf702ce4e1a83b934
    │   ├── 76e821f1b6f0a9709293c3b6b51ed90980b3166b
    │   ├── 82d05b0e688d7ea94675678646c427907419346e
    │   ├── 9d6a233ff6fd5ccb9f76fd99618d73369c52dd3d8222376384d0e601911089e8
    │   ├── a2b5134f4dbc140d9c11f11cba3233099e00af40f262f136c691fb7d38d2194c
    │   └── d008943c017f0092921106440254dbbe00b6a285f7883ec8ba160c3faad88334
    ├── refs
    │   └── main
    └── snapshots
        └── 39593d5650112b4cc580433f6b0435385882d819
            ├── feature_extractor
            │   └── preprocessor_config.json -> ../../../blobs/5294955ff7801083f720b34b55d0f1f51313c5c5
            ├── model_index.json -> ../../blobs/6866dceb3a870b077eb970ecf702ce4e1a83b934
            ├── safety_checker
            │   ├── config.json -> ../../../blobs/5dbd88952e7e521aa665e5052e6db7def3641d03
            │   └── model.safetensors -> ../../../blobs/9d6a233ff6fd5ccb9f76fd99618d73369c52dd3d8222376384d0e601911089e8
            ├── scheduler
            │   └── scheduler_config.json -> ../../../blobs/82d05b0e688d7ea94675678646c427907419346e
            ├── text_encoder
            │   ├── config.json -> ../../../blobs/4d3e873ab5086ad989f407abd50fdce66db8d657
            │   └── model.safetensors -> ../../../blobs/d008943c017f0092921106440254dbbe00b6a285f7883ec8ba160c3faad88334
            ├── tokenizer
            │   ├── merges.txt -> ../../../blobs/76e821f1b6f0a9709293c3b6b51ed90980b3166b
            │   ├── special_tokens_map.json -> ../../../blobs/2c2130b544c0c5a72d5d00da071ba130a9800fb2
            │   ├── tokenizer_config.json -> ../../../blobs/5ba7bf706515bc60487ad0e1816b4929b82542d6
            │   └── vocab.json -> ../../../blobs/469be27c5c010538f845f518c4f5e8574c78f7c8
            ├── unet
            │   ├── config.json -> ../../../blobs/1a02ee8abc93e840ffbcb2d68b66ccbcb74b3ab3
            │   └── diffusion_pytorch_model.safetensors -> ../../../blobs/19da7aaa4b880e59d56843f1fcb4dd9b599c28a1d9d9af7c1143057c8ffae9f1
            └── vae
                ├── config.json -> ../../../blobs/55d78924fee13e4220f24320127c5f16284e13b9
                └── diffusion_pytorch_model.safetensors -> ../../../blobs/a2b5134f4dbc140d9c11f11cba3233099e00af40f262f136c691fb7d38d2194c
```

As one can see many blobs have an identical hash, e.g.:

```
76e821f1b6f0a9709293c3b6b51ed90980b3166
5294955ff7801083f720b34b55d0f1f51313c5c5
...
````

Could it make sense to somehow create a mechanism that does not download files if the exact same hash is already present in a blob of another model repo? 

**Solution**

I wonder whether we can just add a `smart_cache=True` flag to `snapshot_download` that if set, it'll check the `cache_dir` before whether the existing file already exsits in another folder and if yes only references the new file.

I'm wondering whether this is a bit overkill for now though.

It's an issue that originated from discussions with [InvokeAI](https://github.com/invoke-ai/InvokeAI) and is also discussed on diffusers here: https://github.com/huggingface/diffusers/issues/1984

Maybe also cc @julien-c

**Comment by Wauplin** [comment]
Thanks for the detailed feature request here @patrickvonplaten

A few things popped-up in mind when thinking about it:

1. Do you see this problem to be related more to disk usage or to network bandwidth (downloading the files)?
2. TBH, I am not a big fan of having a `smart_cache` flag that would change the cache structure based on its usage or not. With time, it will get harder for the users/for `huggingface_hub` to manage the cache (e.g. _"can I delete this repo from the cache or will it break another cached repo?"_).
We could start to have a "blob index" that keeps track of all blobs and the symlinks between them. But I feel this is error-prone in the future -as any index, it would need to be refreshed from time to time-. The current cache-system is not so bad in the sense that it's efficient for a single repo + it you want to delete it, it's all self-contained.
3. What we could still do is instead of downloading a blob file, we check if we already have it in cache and if yes we _duplicate_ it. It would not solve the disk-usage problem but it seems to be a quick win if you need to test different repos sharing the same file. 
4. If disk usage efficiency _really_ is a problem and if we encounter a lot of cache duplication, maybe we could think of having a single "blobs" folder shared between all repos. It would require:
    1. A script to change existing cache from users
    2. A "blob-index" tracking which repos uses which blobs
    3. Adapt the scan-cache/delete-cache commands (would be straightforward once we have the blob-index)
I'm still a bit reluctant to do that as it would be a major change in users' cache + still not sure there is a lot of demand for such use case.

WDYT? Also interested in @julien-c's opinion since you designed this cache structure (with @LysandreJik I guess?)

**Comment by Wauplin** [comment]
Also 1 thing maybe relevant: when a user do `hf_hub_download` or `snapshot_download`, it return a path to the cached file. It is actually a path to a symlink to the blob file. The user is expected to _read_ the file but never to _write_ in it. If a user do so, its cache will become corrupted, especially if it's the blob file itself that has been edited.

We can't really mitigate this issue and it's at the user's discretion to use the filepath correctly. However in case of a bad usage, only 1 cached repo is impacted. The user can always delete it and restart. If we start to share blobs between repos, this kind of issues can be much more difficult to track down -except deleting the entire cache-. This is also the case if we duplicate blobs between repos (instead of redownloading) - we could end up duplicating a corrupted blob file.

A way of mitigating this is to have a kinda-`sanitize-cache` command that checks every blob file, compute the sha and report if it's not consistent with the blob name (or check blob name/sha consistency just before duplicating it). Wouldn't be too hard to do, but still some extra work.

**Comment by julien-c** [comment]
also would be interested in @coyotte508's opinion as he recently computed the % of LFS files that are duplicated across the whole Hub (and it's not super large). Also cc @Kakulukian 

My hunch would be that this is a bit overkill, but indeed if we do go this way in the future, I wouldn't do this as an option, but always store blobs on a flat level, as @Wauplin says.

Note also that while collisions between SHA256 (what LFS uses) are rare, collisions between SHA1 (what raw git uses, so for non-LFS files) are definitely a thing.

So it would be easy to create a random file that overrides any other file (user still needs to download it to their local cache, of course)

**Comment by julien-c** [comment]
TL;DR: My hunch is that for now, let's monitor the evolution of the % of file duplication across the Hub. If it goes up in the coming months, let's re-assess

**Comment by Wauplin** [comment]
> TL;DR: My hunch is that for now, let's monitor the evolution of the % of file duplication across the Hub. If it goes up in the coming months, let's re-assess

While I agree with this ("let's not care too much until it's really a problem"), let's keep in mind that we are not talking here about % of file duplication across the Hub but % of file duplication between repos that are used by a same roup of users (and in this case, `diffusers` users).

**Comment by julien-c** [comment]
hmm, yes, but the two are bound to be highly correlated, no? (maybe not for a few users, but averaged on all users)

**Comment by Wauplin** [comment]
Yep on average it's the same. For now it seems that it's mostly impacting only (some?) `diffusers` users.

In any case, @patrickvonplaten unless you think there is a huge need for such an optimization, I don't think we will change something in the near future.

**Comment by keturn** [comment]
>  For now it seems that it's mostly impacting only (some?) diffusers users.

Is this pattern of "construct a pipeline that makes use of multiple models and store all those models under a single repo" a pattern that's unique to diffusers?

**Comment by Wauplin** [comment]
> Is this pattern of "construct a pipeline that makes use of multiple models and store all those models under a single repo" a pattern that's unique to diffusers?

At least for what I am aware of, yes it is. Let's ping @sgugger to be sure it's not the case in `transformers` ? At least this optimization has not been suggested so far.

(@sgugger TL;DR of this topic: `diffusers` have some pipelines re-using a lot the same files between repos. In the current cache-system it means same files are duplicated (once per repo) -apparently there could be a 20% gain of disk usage in some cases-. For now we will not proceed further in this direction to keep repos cache "self-contained")

**Comment by sgugger** [comment]
We don't use `snapshot_download` at all in Transformers, so the problem doesn't appear :-)

**Comment by Wauplin** [comment]
Actually the question is more around the cache-system itself so both `hf_hub_download` and `snapshot_download`.

**Comment by sgugger** [comment]
We also don't re-use the same large files between repos. Some tokenizer files are duplicated but they're not too big so it's not a problem on our end.

**Comment by patrickvonplaten** [comment]
It's quite unique to `diffusers` (for now at least). If many more "composite" models (VQ-Tokenization + generation) would find their way into `transformers`, then it could become interesting there as well, but that's not really the case and doesn't seem to happen anytime soon
